lint-consul-retry.go

```diff
@@ -253,16 +257,46 @@ func usesRequire(fun ast.Expr) bool {
 	return false
 }
 
-// usesParam checks if param is first in a call expression
-func usesParam(param string, ce *ast.CallExpr) bool {
-	// t is always first arg to require when not using require.New
-	firstArg, ok := ce.Args[0].(*ast.Ident)
-	if !ok {
-		return false
-	}
-	if firstArg.Name == param {
-		return true
+func usesTestingT(ce *ast.CallExpr) bool {
+	for _, raw := range ce.Args {
+		arg, ok := raw.(*ast.Ident)
+		if !ok {
+			continue
+		}
+
+		if arg.Obj == nil || arg.Obj.Kind != ast.Var {
+			continue
+		}
+
+		field, ok := arg.Obj.Decl.(*ast.Field)
+		if !ok {
+			continue
+		}
+
+		ptr, ok := field.Type.(*ast.StarExpr)
+		if !ok {
+			continue
+		}
+
+		sel, ok := ptr.X.(*ast.SelectorExpr)
+		if !ok {
+			continue
+		}
+
+		if sel.Sel.Name != "T" {
+			continue
+		}
+
+		pkg, ok := sel.X.(*ast.Ident)
+		if !ok {
+			continue
+		}
+
+		if pkg.Name == "testing" {
```

Comment:
What about `require.TestingT`, or any interface type that is satisfied by `testing.T` and where we pass a testing.T type to the func? Would we be able to catch that with this?

Reply:
Using a concrete `testing.T` is always wrong within a retry. The trouble with the `testing.TB` interface and other concrete types that may be used that implement those interfaces is that we cannot know for certain whether the usage is bad. `require.TestingT` is a good example. `require.Equal` (and other assertion methods) will pass the `TestingT` parameter off to `assert.Equal`. This usage of `require.TestingT` as a parameter to another function clearly correct. 

If we had more complicated static analysis we might be able to detect that a require.TestingT was created prior to the retry invocation and thus shouldn't be used either. While it would provide even more protections it gets into the realm of probably not worth it. The utility of this linter is that its relatively simple yet can catch a pretty broad swathe of errors. Its definitely not perfect but without significant investments of time, its probably the best we can do for now.